### PR TITLE
virt: Add new virt_secret module

### PR DESCRIPTION
**Pull request by theo-o**
##### SUMMARY
- Used to manage Libvirt secrets as per [https://libvirt.org/formatsecret.html](https://libvirt.org/formatsecret.html)

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
virt_secret

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/cloud/misc/virt_secret.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    short_description: Manages Libvirt secrets
+    description:
+        - Manages Libvirt secrets.
+    version_added: "2.8"
+    author: "Theo Ouzhinski (@theo-o)"
+    options:
+        uuid:
+            description:
+                - The UUID of the secret being managed.
+                - Must be in the format specified by RFC4122.
```
Comment (jamescassell) [inline lib/ansible/modules/cloud/misc/virt_secret.py]: The docs page you linked says it doesn't have to be in that format.

File: lib/ansible/modules/cloud/misc/virt_secret.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+                - Required when I(state=present).
+            choices:
+                - volume
+                - ceph
+                - iscsi
+                - tls
+        usage_element:
+            description:
+                - Required when usage is defined.
+                - See L(the Libvirt docs, https://libvirt.org/formatsecret.html) for more information.
```
Comment (jamescassell) [inline lib/ansible/modules/cloud/misc/virt_secret.py]: I read the docs page, but it's not at all clear what this usage_element is for, nor do the examples provided below give me a good idea.

File: lib/ansible/modules/cloud/misc/virt_secret.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    usage: ceph
+    usage_element: client.libvirt secret
+    private: no
+    ephemeral: no
+    state: present
+
+- name: Define a Ceph secret with a value and description (with random UUID)
+  virt_secret:
+    secret: AQBHCbtT6APDHhAA5W00cBchwkQjh3dkKsyPjw==
+    usage: ceph
```
Comment (jamescassell) [inline lib/ansible/modules/cloud/misc/virt_secret.py]: Would be good to also see usage other than ceph.

File: lib/ansible/modules/cloud/misc/virt_secret.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                - ceph
+                - iscsi
+                - tls
+        usage_element:
+            description:
+                - Required when usage is defined.
+                - See L(the Libvirt docs, https://libvirt.org/formatsecret.html) for more information.
+        description:
+            description:
+                - A human-readable description of the secret''s purpose
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_secret.py]:
```suggestion
                - "A human-readable description of the secret's purpose."
```

File: lib/ansible/modules/cloud/misc/virt_secret.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+            secret=dict(type='str'),
+            usage=dict(type='str', choices=USAGE_CHOICES),
+            usage_element=dict(type='str'),
+            description=dict(type='str'),
+            ephemeral=dict(type='bool'),
+            private=dict(type='bool'),
+            state=dict(type='str', choices=STATE_CHOICES, required=True)),
+        supports_check_mode=True)
+
+    binary_path = module.get_bin_path('virsh')
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_secret.py]:
```suggestion
    binary_path = module.get_bin_path('virsh', required=True)
```

File: lib/ansible/modules/cloud/misc/virt_secret.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+    ephemeral = module.params.get('ephemeral', None)
+    private = module.params.get('private', None)
+    state = module.params.get('state', None)
+
+    global result
+    result = dict(
+        changed=False,
+        uuid='')
+
+    if not binary_path:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/misc/virt_secret.py]: This is not required if you use `required=True` in `get_bin_path` API.

**Comment by theo-o** [comment]
Thanks for your review @Akasurde and @jamescassell.  Is there anything I need to do in order to advance this PR?

**Review by Akasurde — APPROVED** [review]
LGTM

**Comment by jamescassell** [comment]
Two things:
1. Why force the `uuid` param to be a valid UUID when libvirt does not enforce that?
2. Would be good to see integration tests.

**Pull request closed without merging**